lib/puppet/pops/types/type_factory.rb

```diff
@@ -351,6 +351,22 @@ def self.hash_of_data()
     type
   end
 
+  # Produces a type for NotUndef[T]
+  # The given 'inst_type' can be a string in which case it will be converted into
+  # the type String[inst_type].
+  #
+  # @param inst_type [Type,String] the type to qualify
+  # @return [Puppet::Pops::Types::PNotUndefType] the NotUndef type
+  #
+  # @api public
+  #
+  def self.not_undef(inst_type = nil)
+    type = Types::PNotUndefType.new()
+    inst_type = string(inst_type) if inst_type.instance_of?(String)
```

Comment:
We typically use .is_a? everywhere instead of instance_of? which requires that the object is an instance of exactly that class, and not a subclass. This can something screw things up when running tests and doing mocking.